src/math/quadtree.cpp

```diff
@@ -0,0 +1,137 @@
+//  SuperTux -  A Jump'n Run
+//  Copyright (C) 2018 Christian Hagemeier <christian@hagemeier.ch>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "math/quadtree.hpp"
+
+QuadTree::QuadTree(Rectf r, int level):
+  m_level(level),
+  m_objects(),
+  m_onlyhere(),
+  m_bounds(r),
+  next()
+  {
+    for(int i = 0;i<4;i++)
+      next[i] = NULL;
+  }
+
+void QuadTree::insert(MovingObject * obj)
+{
+  if(next[0] != NULL)
+  {
+      int nindex = getIndex(obj);
+      if(nindex != -1)
+      {
+        next[nindex]->insert(obj);
+        return;
+      }else{
+        m_onlyhere.push_back(obj);
+      }
+  }
+  m_objects.push_back(obj);
+  // Is our capacity fully used?
+  if(m_objects.size() > MAX_OBJECTS && m_level < MAX_LAYERS)
+  {
+    split();
+    // Insert every object in respective quadrant.
+    for(auto it = m_objects.begin();it != m_objects.end(); it++)
```

Comment:
Can you use the shorter
```
for(const auto& nobj : m_objects)
```
or something to that effect?

The thing is: I can never remember whether the above long way creates copies or not.